### PR TITLE
[foreach][Adam] Minimize use of intermediates to decrease peak memory

### PR DESCRIPTION
**Pull request by janeyx99**
Starts addressing https://github.com/pytorch/pytorch/issues/97712 by
- Minimizing intermediates usage for foreach Adam
- Document the extra memory usage 
- Add comments within the code for clarity now that we reuse intermediates
- Add tests
- Did some refactoring

Next steps involve doing this for all other foreach implementations. Note that even after this change, foreach mem usage will be higher than forloop due to the fact that we have a minimum budget of 1 intermediate (to not muddle the input values) and the intermediate will be larger. For capturable, the memory usage is higher due to moving more tensors to CUDA.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104780
* #104779
* #104669

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -473,42 +477,51 @@ def _multi_tensor_adam(params: List[Tensor],
...
+            torch._foreach_div_(bias_correction1, lr)
+            torch._foreach_reciprocal_(bias_correction1)
 
-            bias_correction2_sqrt = torch._foreach_sqrt(bias_correction2)
+            torch._foreach_sqrt_(bias_correction2)
+
+            # Re-assign for clarity as we maintain minimal intermediates: we'll have
+            # step_size = - lr / (1 - beta1 ^ t) where t = num_steps
+            # bias_correction2 = 1 - beta2 ^ t
+            step_size = bias_correction1
```
Comment (albanD) [inline torch/optim/adam.py:501]: That sounds good. Just mentioning as well that you can do things like `step_size = torch._foreach_reciprocal_(bias_correction1)` or `bias_correction2_sqrt = torch._foreach_sqrt_(bias_correction2)` if you prefer.

File: torch/optim/adam.py [hunk 2]
```diff
@@ -473,42 +477,51 @@ def _multi_tensor_adam(params: List[Tensor],
...
             else:
                 exp_avg_sq_sqrt = torch._foreach_sqrt(device_exp_avg_sqs)
-                torch._foreach_div_(exp_avg_sq_sqrt, torch._foreach_mul(bias_correction2_sqrt, step_size))
-                eps_over_step_size = torch._foreach_div(step_size, eps)
-                torch._foreach_reciprocal_(eps_over_step_size)
-                denom = torch._foreach_add(exp_avg_sq_sqrt, eps_over_step_size)
 
-            torch._foreach_addcdiv_(device_params, device_exp_avgs, denom)
+            # Folds in (admittedly ugly) 1-elem step_size math here to avoid extra param-set-sized read+write
+            # (can't fold it into addcdiv_ below because addcdiv_ requires value is a Number, not a Tensor)
```
Comment (albanD) [inline torch/optim/adam.py:514]: I don't think this is true.

**Review by albanD — APPROVED** [review]
Small stylistic nits but sound good otherwise.

File: torch/optim/adam.py (hunk 1 above)
Comment (janeyx99) [reply torch/optim/adam.py:501]: Nah, I'm p sure those return nothing. I had something like that before and they returned None. 
image

Comment (albanD) [reply torch/optim/adam.py:501]: What? :o  All our other inplace ops return the modified arg... This is very weird...

Comment (janeyx99) [reply torch/optim/adam.py:501]: "foreach support is beta" i guess? I can make another issue tracking this when we wanna make foreach legit

File: torch/optim/adam.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/adam.py]: Rewrote to be cleaner...though I still didn't need addcdiv 🤔

File: torch/optim/adam.py (hunk 1 above)
Comment (crcrpar) [reply torch/optim/adam.py:501]: just so you know IIRC the current torchgen doesn't allow inplace foreach functions to return a list of Tensors (I don't remember the exact lines off the top of my head)

Comment (janeyx99) [reply torch/optim/adam.py:501]: ioh that would make sense

Comment (crcrpar) [reply torch/optim/adam.py:501]: I re-grepped it:
https://github.com/pytorch/pytorch/blob/d8cb80e3827226eeaafd6a007018112301f89d41/torchgen/model.py#L1411-L1428

**Pull request closed without merging**